core/src/main/scala/filodb.core/downsample/ChunkDownsampler.scala

```diff
@@ -0,0 +1,272 @@
+    */
+  def downsampleChunk(part: TimeSeriesPartition,
+                      chunkset: ChunkSetInfo,
+                      startRow: Int,
+                      endRow: Int): Double
+}
+
+trait TimestampChunkDownsampler extends ChunkDownsampler {
+  override val colType: ColumnType = ColumnType.TimestampColumn
+
+  /**
+    * Downsamples Chunk using timestamp column Ids configured and emit long value
+    * @param part Time series partition to extract data from
+    * @param chunkset The chunksetInfo that needs to be downsampled
+    * @param startRow The start row number for the downsample period (inclusive)
+    * @param endRow The end row number for the downsample period (inclusive)
+    * @return downsampled value to emit
+    */
+  def downsampleChunk(part: TimeSeriesPartition,
+                      chunkset: ChunkSetInfo,
+                      startRow: Int,
+                      endRow: Int): Long
+}
+
+/**
+  * Downsamples by calculating sum of values in one column
+  */
+case class SumDownsampler(override val colIds: Array[Int]) extends DoubleChunkDownsampler {
+  require(colIds.length == 1, s"Sum downsample requires only one column. Got ${colIds.length}")
+  override val name: String = "dSum"
+  override def downsampleChunk(part: TimeSeriesPartition,
+                               chunkset: ChunkSetInfo,
+                               startRow: Int,
+                               endRow: Int): Double = {
+
+    val vecPtr = chunkset.vectorPtr(colIds(0))
+    val colReader = part.chunkReader(colIds(0), vecPtr)
+    colReader.asDoubleReader.sum(vecPtr, startRow, endRow)
+  }
+}
+
+/**
+  * Downsamples by calculating count of values in one column
+  */
+case class CountDownsampler(override val colIds: Array[Int]) extends DoubleChunkDownsampler {
+  require(colIds.length == 1, s"Count downsample requires only one column. Got ${colIds.length}")
+  override val name: String = "dCount"
+  override def downsampleChunk(part: TimeSeriesPartition,
+                               chunkset: ChunkSetInfo,
+                               startRow: Int,
+                               endRow: Int): Double = {
+
+    val vecPtr = chunkset.vectorPtr(colIds(0))
+    val colReader = part.chunkReader(colIds(0), vecPtr)
+    colReader.asDoubleReader.count(vecPtr, startRow, endRow)
+  }
+}
+
+/**
+  * Downsamples by calculating min of values in one column
+  */
+case class MinDownsampler(override val colIds: Array[Int]) extends DoubleChunkDownsampler {
+  require(colIds.length == 1, s"Min downsample requires only one column. Got ${colIds.length}")
+  override val name: String = "dMin"
+  override def downsampleChunk(part: TimeSeriesPartition,
+                               chunkset: ChunkSetInfo,
+                               startRow: Int,
+                               endRow: Int): Double = {
+
+    val vecPtr = chunkset.vectorPtr(colIds(0))
+    val colReader = part.chunkReader(colIds(0), vecPtr)
+
+    var min = Double.MaxValue
```

Comment:
This code overlaps with the code in the ChunkedMinRangeFunction thing (forgot exact name).  Maybe we should move it to a more common place, like in the DoubleReader.

Reply:
I tried this, but am possibly missing some information to cover all implementations. Can we leave the refactor out to a subsequent PR? I don't want to digress too much from the scope of this PR.